### ISSUE TITLE
File permissions in Helm operator container

### ISSUE THREAD

**Issue by JohnStrunk** [operator-framework/operator-sdk#1206]
## Bug Report

**What did you do?**
I created and built the scaffolding for a Helm operator, but the file permissions inside the container are such that the chart and `watches.yaml` are unreadable.

Steps:
```
$ operator-sdk new appservice --skip-git-init --api-version app.example.com/v1alpha1 --kind AppService --type helm
INFO[0000] Creating new Helm operator 'appservice'.     
INFO[0000] Create helm-charts/appservice/               
INFO[0000] Create build/Dockerfile                      
INFO[0000] Create watches.yaml                          
INFO[0000] Create deploy/service_account.yaml           
INFO[0000] Create deploy/role.yaml                      
INFO[0000] Create deploy/role_binding.yaml              
INFO[0000] Create deploy/operator.yaml                  
INFO[0000] Create deploy/crds/app_v1alpha1_appservice_crd.yaml 
INFO[0000] Create deploy/crds/app_v1alpha1_appservice_cr.yaml 
INFO[0000] Project creation complete.                   

$ cd appservice/
$ operator-sdk build appservice
INFO[0000] Building Docker image appservice             
Sending build context to Docker daemon  31.23kB
Step 1/3 : FROM quay.io/operator-framework/helm-operator:v0.5.0
 ---> f0d56774da3e
Step 2/3 : COPY helm-charts/ ${HOME}/helm-charts/
 ---> e4b5501c936d
Step 3/3 : COPY watches.yaml ${HOME}/watches.yaml
 ---> 102b9945d010
Successfully built 102b9945d010
Successfully tagged appservice:latest
INFO[0000] Operator build complete.
```

Taking a look inside the resulting image:

```
$ docker run -it --rm --entrypoint /bin/sh appservice
/ $ ls /opt/helm/ -al
total 32
drwxrwxr-x    1 1001     root          4096 Mar 13 20:22 .
drwxr-xr-x    1 root     root          4096 Feb  8 20:11 ..
-rw-------    1 1001     root            32 Mar 13 20:22 .ash_history
drwxr-xr-x    3 root     root          4096 Mar 13 20:22 helm-charts
-rw-------    1 root     root           110 Mar 13 20:21 watches.yaml
/ $ ls -al /opt/helm/helm-charts/appservice/
ls: can't open '/opt/helm/helm-charts/appservice/': Permission denied
total 0
```

**What did you expect to see?**
The chart & watches file should be readable by the userid that is running in the container.

**What did you see instead? Under which circumstances?**
As created, the added files are only readable by root (uid=0), making them inaccessible.

Picking through the sdk a bit, I suspect this is the result of the permissions on the files created by the scaffolding, combined w/ the creator's umask. The `COPY` lines on the Dockerfile just bring the file permissions across and set the uid/gid to 0. Adding a `RUN` to the Dockerfile that does a `chmod` results in permission denied because the base image sets the UID to non-root here: https://github.com/operator-framework/operator-sdk/blob/3a8d517f5422ad7171c9174ae1a98e07c367482d/test/helm-operator/Dockerfile#L16

I suspect this problem only appears where users have a restrictive umask (0066 in my case).

**Environment**
* operator-sdk version:

```
$ operator-sdk --version
operator-sdk version v0.5.0
```

* Kubernetes version information:

N/A

* Kubernetes cluster kind: 

* Are you writing your operator in ansible, helm, or go?

helm

**Possible Solution**

I think the easiest fix is to move the `USER ${USER_UID}` from the base image to being part of the final operator's Dockerfile. This should permit a `RUN chmod a+r -R ${HOME}` to be added after the COPY lines in the operator Dockerfile.

Alternatively, having the `operator-sdk new` set the initial permissions to a+r will fix the files that are a part of the scaffolding, but it doesn't help with the chart template files created by the user as they develop their helm chart.

**Additional context**

**Comment by joelanford** [comment]
@JohnStrunk Thanks for the bug report. I was able to reproduce and agree this is something we should be able fix from the operator-sdk.

Which version of docker are you using? In 17.09+, the `COPY` command supports a `--chown` flag. I'm wondering if you can use this in your Dockerfile:

```Dockerfile
FROM quay.io/operator-framework/helm-operator:v0.5.0

COPY --chown=1001:0 helm-charts/ ${HOME}/helm-charts/
COPY --chown=1001:0 watches.yaml ${HOME}/watches.yaml

RUN find ${HOME} -type d -exec chmod g+rx {} \; && \
    find ${HOME} -type f -exec chmod g+r {} \;
```

If that works, maybe we should consider using that in our Dockerfile template and bumping our Docker version requirement to 17.09+.

I don't think we should publish a helm-operator base image that runs as `root` by default.

**Comment by JohnStrunk** [comment]
Those changes worked for me, though I'm using `go+rx` & `go+r` in the event the environment forces a non-zero gid.

I'm a bit unhappy about having to hardcode the uid of 1001 in the chown since that requires knowledge of how the base container is built. I tried `${USER_UID}` (which seems a bit less ugly), but it looks like docker won't substitute the variable in the chown option.

Thanks for looking into it. :+1:

**Comment by joelanford** [comment]
> I'm a bit unhappy about having to hardcode the uid of 1001 in the chown since that requires knowledge of how the base container is built. I tried ${USER_UID} (which seems a bit less ugly), but it looks like docker won't substitute the variable in the chown option.

I agree. Would it be more palatable if we created a user entry in `/etc/passwd` for uid 1001, so that we could make it something like `--chown=helm-operator:root`?

That would still require knowledge about which users are present in the base container, but that seems better than a uid.

**Comment by JohnStrunk** [comment]
I agree, that looks better.

As for the name, when the container is actually run, `entrypoint` adds user "helm" to the password file. It seems like the build-time and run-time usernames should be noticeably different since the uids are unlikely to coincide in practice. Perhaps: `helm-builder` or `container-builder` to distinguish it from the `helm` user at run-time.

**Comment by feli5** [comment]
After some dig, I find it works with root user.
And there is no user with id 1001

# docker exec -it 5949f84398fb bash
bash-4.2$ whoami
whoami: cannot find name for user ID 1001

```

# docker exec -it -u root 5949f84398fb bash
bash-4.2# mkdir /var/run/secrets/kubernetes.io/serviceaccount
```
Then, copy secrets from host

```
bash-4.2# export WATCH_NAMESPACE=default
bash-4.2# export POD_NAME=app-operator-75d9c94849-6mqns
bash-4.2# export OPERATOR_NAME=app-operator
bash-4.2# export KUBERNETES_PORT_443_TCP_PROTO=tcp
bash-4.2# export KUBERNETES_PORT_443_TCP_PORT=443
bash-4.2# export KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1
bash-4.2# export KUBERNETES_SERVICE_HOST=10.96.0.1
bash-4.2# export KUBERNETES_SERVICE_PORT=443
bash-4.2# export KUBERNETES_SERVICE_PORT_HTTPS=443
bash-4.2# export KUBERNETES_PORT=tcp://10.96.0.1:443
bash-4.2# export KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443
bash-4.2# export OPERATOR=/usr/local/bin/app-operator
bash-4.2# export container=oci
bash-4.2# export PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin^C
bash-4.2# export USER_UID=1001
bash-4.2# export USER_NAME=app-operator
bash-4.2# app-operator
{"level":"info","ts":1559399403.634943,"logger":"cmd","msg":"Go Version: go1.12.5"}
{"level":"info","ts":1559399403.6349921,"logger":"cmd","msg":"Go OS/Arch: linux/amd64"}
{"level":"info","ts":1559399403.6349971,"logger":"cmd","msg":"Version of operator-sdk: v0.8.0"}
{"level":"info","ts":1559399403.6351745,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1559399403.7024112,"logger":"leader","msg":"No pre-existing lock was found."}
{"level":"info","ts":1559399403.713179,"logger":"leader","msg":"Became the leader."}
{"level":"info","ts":1559399403.7495632,"logger":"cmd","msg":"Registering Components."}
{"level":"info","ts":1559399403.7497835,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"appservice-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1559399403.749956,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"appservice-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1559399403.8155308,"logger":"metrics","msg":"Metrics Service object created","Service.Name":"app-operator","Service.Namespace":"default"}
{"level":"info","ts":1559399403.8155599,"logger":"cmd","msg":"Starting the Cmd."}
{"level":"info","ts":1559399403.915885,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"appservice-controller"}
{"level":"info","ts":1559399404.0160387,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"appservice-controller","worker count":1}

```



```

**Comment by estroz** [comment]
@JohnStrunk are you still experiencing this issue? If so, feel free to re-open.

**Comment by camilamacedo86** [comment]
Just to track here. 
I checked it with the master branch and then, I was unable to reproduce the scenario. 

```
$ operator-sdk new appservice --api-version app.example.com/v1alpha1 --kind AppService --type helm
INFO[0000] Creating new Helm operator 'appservice'.     
INFO[0000] Created helm-charts/appservice               
INFO[0000] Generating RBAC rules                        
WARN[0000] The RBAC rules generated in deploy/role.yaml are based on the chart's default manifest. Some rules may be missing for resources that are only enabled with custom values, and some existing rules may be overly broad. Double check the rules generated in deploy/role.yaml to ensure they meet the operator's permission requirements. 
INFO[0000] Created build/Dockerfile                     
INFO[0000] Created watches.yaml                         
INFO[0000] Created deploy/service_account.yaml          
INFO[0000] Created deploy/role.yaml                     
INFO[0000] Created deploy/role_binding.yaml             
INFO[0000] Created deploy/operator.yaml                 
INFO[0000] Created deploy/crds/app.example.com_v1alpha1_appservice_cr.yaml 
INFO[0000] Generated CustomResourceDefinition manifests. 
INFO[0000] Project creation complete.                   
camilamacedo@Camilas-MacBook-Pro ~/go/src $ cd appservice/
camilamacedo@Camilas-MacBook-Pro ~/go/src/appservice $ operator-sdk build appservice
INFO[0000] Building OCI image appservice                
Sending build context to Docker daemon  32.26kB
Step 1/3 : FROM quay.io/operator-framework/helm-operator:v0.15.0
v0.15.0: Pulling from operator-framework/helm-operator
7fe5fcc0340e: Pull complete 
086296bbdfc7: Pull complete 
ac9a30764ebf: Pull complete 
853416420bb2: Pull complete 
3267895c24b5: Pull complete 
Digest: sha256:986b7fb69cfd1192091b0656af88e0aac48e3d96e35a7ef8705a6b84cf2dedbd
Status: Downloaded newer image for quay.io/operator-framework/helm-operator:v0.15.0
 ---> 89624e945eea
Step 2/3 : COPY watches.yaml ${HOME}/watches.yaml
 ---> 5d682a30024c
Step 3/3 : COPY helm-charts/ ${HOME}/helm-charts/
 ---> c0339446b918
Successfully built c0339446b918
Successfully tagged appservice:latest
INFO[0015] Operator build complete.                     
camilamacedo@Camilas-MacBook-Pro ~/go/src/appservice $ docker run -it --rm --entrypoint /bin/sh appservice
sh-4.4$ ls -al /opt/helm/helm-charts/appservice/
total 28
drwxr-xr-x 4 root root 4096 Mar 11 09:29 .
drwxr-xr-x 3 root root 4096 Mar 11 09:30 ..
-rw-r--r-- 1 root root  342 Mar 11 09:29 .helmignore
-rw-r--r-- 1 root root  908 Mar 11 09:29 Chart.yaml
drwxr-xr-x 2 root root 4096 Mar 11 09:29 charts
drwxr-xr-x 3 root root 4096 Mar 11 09:29 templates
-rw-r--r-- 1 root root 1493 Mar 11 09:29 values.yaml
sh-4.4$ ls /opt/helm/ -al
total 16
drwxrwxr-x 1 1001 root 4096 Mar 11 09:30 .
drwxr-xr-x 1 root root 4096 Jan 22 22:53 ..
drwxr-xr-x 3 root root 4096 Mar 11 09:30 helm-charts
-rw-r--r-- 1 root root  100 Mar 11 09:29 watches.yaml
sh-4.4$ 

```
